runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -56,181 +57,439 @@ extern "C" {
-    // call site, which is located by the "return address" field of the current stack frame, contains information necessary to locate the entry points of both the calling and called functions and, if required,
-    // information about floating-point parameters passed. This is done by encoding information in a no-op instruction at the return point.
-    // This is encoded using a 2–byte NOPR instruction, like the BCR below.
-    thunkMem += BCR(thunkMem, 0, 0);
-    // Upon returning copy back the structure, then restore the stack pointer.
-    thunkMem += LMG(thunkMem, R6, R15, R4, offsetToRegSaveArea + thunkFrameSize); // 1808+256
-    thunkMem += LAY(thunkMem, R4, 0, R4, thunkFrameSize);
-    // Then branch back
-    thunkMem += BC(thunkMem, 0xF, 0, R7, 2);
+	// Now generate the branch to the helper
+	thunkMem += BASR(thunkMem, R7, R6);
 
-    // Finish up before returning
-    vmFuncs->doneUpcallThunkGeneration(metaData, metaData->thunkAddress);
-    metaData->functionPtr[0] = reinterpret_cast<UDATA>((reinterpret_cast<uint64_t *>(metaData->upCallCommonDispatcher)));
-    metaData->functionPtr[1] = reinterpret_cast<UDATA>(metaData->thunkAddress);
+	// Mandatory NOP instruction
+	// With XPLINK, the function entry point address is not always passed to the called function. To allow Language Environment and other tools to find the entry point of the currently executing routine, every
+	// call site, which is located by the "return address" field of the current stack frame, contains information necessary to locate the entry points of both the calling and called functions and, if required,
+	// information about floating-point parameters passed. This is done by encoding information in a no-op instruction at the return point.
+	// This is encoded using a 2 byte NOPR instruction, like the BCR below.
+	thunkMem += BCR(thunkMem, 0, 0);
 
-   return (void *)(&(metaData->functionPtr));
+	// Handle struct return type
+	if ((returnType & J9_FFI_UPCALL_SIG_TYPE_MASK) == J9_FFI_UPCALL_SIG_TYPE_STRUCT) {
+		// After returning we load the return struct value into registers or memory (if applicable)
+		if (returnType == J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_SP && returnValueSize == 2 * sizeof(float)) {
+			// In this case, GPR3 has an address. We need to load the first 4 bytes at this address into FPR0 and the next 4 bytes into FPR2
+			thunkMem += LE(thunkMem, FPR0, 0, R3, 0);
+			thunkMem += LE(thunkMem, FPR2, 0, R3, 4);
+		}
+		else if (returnType == J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_DP && returnValueSize == 2 * sizeof(double)) {
+			thunkMem += LD(thunkMem, FPR0, 0, R3, 0);
+			thunkMem += LD(thunkMem, FPR2, 0, R3, 8);
+		}
+		else if (returnValueSize > 0 && returnValueSize <= 9) {
+			thunkMem += LG(thunkMem, R1, 0, R3, 0);
+		}
+		else if (returnValueSize > 8 && returnValueSize <= 16) {
```

Comment:
Isn't this redundant ? i.e, if returnValueSize not <= 8 then only we will reach here right ?